### PR TITLE
perf: use std::nth_element instead of full sort when pruning bounded caches

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
## Issue being fixed or feature implemented

Two bounded caches do strictly more work than they need to when they evict:

- `unordered_limitedmap::prune()` (`src/limitedmap.h`) collects an iterator to every entry in the map and fully sorts that vector ascending by mapped value, then throws away everything except the `tooMuch = size - nMaxSize` smallest entries.
- `unordered_lru_cache::truncate_if_needed()` (`src/unordered_lru_cache.h`) does the same thing with a full descending sort by access counter, then erases everything past index `maxSize`.

In both cases the total order is discarded immediately. All that is actually needed is a partition around the eviction boundary, so the sort is `O(n log n)` work where `O(n)` suffices. This matters because these caches are not small: `unordered_lru_cache` instances in the tree are sized at 10000 (`islockCache`, `txidCache`, `outpointCache`) and 30000 (`hasSigForIdCache`, `hasSigForSessionCache`, `hasSigForHashCache`), and truncation runs on the insert path.

This came out of review of https://github.com/dashpay/dash/pull/7482, which also touches `src/limitedmap.h`. This PR is deliberately scoped so the two can merge in either order without conflicting: it only rewrites the body of `prune()`, leaves the constructor and the `max_size()` / prune-threshold accessors alone, and does not touch `src/test/limitedmap_tests.cpp` at all. The existing limitedmap tests already pin the prune membership semantics that must be preserved.

## What was done?

Replaced `std::sort` with `std::nth_element` in both eviction paths, using the same comparator in each case:

- `unordered_limitedmap::prune()`: `tooMuch` is now computed before the partition, and `std::nth_element(begin, begin + tooMuch, end, cmp)` places the `tooMuch` smallest entries in the leading range. The subsequent `resize(tooMuch)` and erase loop are unchanged.
- `unordered_lru_cache::truncate_if_needed()`: `std::nth_element(vec.begin(), vec.begin() + maxSize, vec.end(), cmp)` leaves the `maxSize` most recently accessed entries in the leading range, and the existing loop erases everything from `maxSize` onward.

This is behaviour preserving for the *set* of evicted entries. The relative order within the evicted group and within the retained group was never observable, and ties between equal values were already broken arbitrarily because `std::sort` is not a stable sort.

Two supporting changes:

- Added `assert(truncateThreshold >= maxSize)` to the `unordered_lru_cache` constructor. `vec.begin() + maxSize` is only a valid iterator because truncation runs only when `size > truncateThreshold`, so that invariant now has to hold rather than merely happening to hold. Every instantiation in `src/` passes only `maxSize` and therefore gets the default `truncateThreshold = 2 * maxSize`, so no caller is affected. `unordered_limitedmap` already has the equivalent `assert(nPruneAfterSize >= nMaxSize)`.
- Added `src/test/unordered_lru_cache_tests.cpp`. The class previously had no unit test coverage at all.

## How Has This Been Tested?

Built on macOS arm64 (aarch64-apple-darwin, clang, `--enable-debug --enable-crash-hooks --enable-stacktraces --without-gui --enable-tests`), against the `depends` prefix.

- `make -C src test/test_dash` at both commits of this branch, to confirm each commit builds on its own.
- `./src/test/test_dash --run_test=limitedmap_tests,unordered_lru_cache_tests` — 6 test cases, no errors.
- `./src/test/test_dash` (full unit test suite) — 770 test cases, no errors. This is the check that the new constructor assert does not fire for any existing cache instantiation.
- `test/lint/lint-whitespace.py`, `test/lint/lint-git-commit-check.py` (with `COMMIT_RANGE` set to this branch), `test/lint/lint-tests.py`, `test/lint/lint-includes.py`, `test/lint/lint-include-guards.py` — all clean.

The new tests cover: no truncation while the size is at or below the truncate threshold (which also pins the default threshold at `2 * max_size`); truncation down to exactly `max_size`, keeping the most recently accessed entries and leaving their values intact; an explicitly supplied truncate threshold; and that both `get()` and `exists()` refresh an entry's recency so an otherwise-oldest entry survives truncation.

Not run: functional tests, since no behaviour reachable from RPC or P2P changes.

## Breaking Changes

None. Both changes preserve the set of entries evicted, and no consensus, network, or RPC behaviour is affected.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [ ] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

**Comment by thepastaclaw** [comment]
🔍 Review in progress — actively reviewing now (commit b498b678)
_Stage: Sonnet review + final verification_
_ETA: complete ~01:05 UTC (median 19m across 30 recent reviews)_
_Running 14m · Last checked: 2026-08-02 01:00 UTC_

**Review by thepastaclaw — COMMENTED** [review]
## Final validation — Codex + Sonnet

The bounded-cache eviction changes are behavior-preserving and the added unit tests cover threshold and recency semantics. One repository-maintenance omission leaves the new Dash-specific test source outside the file set used by Dash cppcheck and differential formatting checks; there are no blocking correctness issues.

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — dash-core-commit-history (completed)
- Verifier: `gpt-5.6-sol` — verifier (fallback)
- Sonnet reviewers: `claude-sonnet-5` — general (completed), `claude-sonnet-5` — dash-core-commit-history (failed), `claude-sonnet-5` — dash-core-commit-history (completed)

🟡 1 suggestion(s)

_1 additional finding(s) omitted (not in diff)._

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `test/util/data/non-backported.txt`:
- [SUGGESTION] test/util/data/non-backported.txt:68-69: Register the new Dash-specific test for lint checks
  `src/test/unordered_lru_cache_tests.cpp` is newly authored Dash-specific code, but this list does not include it. Both `test/lint/lint-cppcheck-dash.py` and the differential clang-format workflow derive their Dash-specific file sets from this list, so the new test is silently excluded from those checks. Add its exact path alongside the other unit-test entries.
```
</details>

File: src/unordered_lru_cache.h [hunk 1]
```diff
@@ -29,6 +29,7 @@ class unordered_lru_cache
     {
         // either specify maxSize through template arguments or the constructor and fail otherwise
         assert(_maxSize != 0);
+        assert(truncateThreshold >= maxSize);
     }
```
Comment (thepastaclaw) [inline src/unordered_lru_cache.h:34]: **🔴 Blocking: Preserve smaller truncation thresholds**

The constructor previously accepted configurations such as `cache(10, 5)`. Although the old implementation began sorting after the sixth insertion, it did not erase anything until the size exceeded `maxSize`, so the configuration remained safe and effectively behaved as if the threshold were 10. The new assertion rejects it only in assertion-enabled builds; with `NDEBUG`, the sixth insertion reaches `std::nth_element(vec.begin(), vec.begin() + 10, vec.end(), ...)` with a six-element vector, forming an out-of-range iterator and invoking undefined behavior. Clamp the effective threshold to `maxSize` instead of relying on a debug-only assertion, and add a regression test for a threshold below the maximum.

```suggestion
    explicit unordered_lru_cache(size_t _maxSize = MaxSize, size_t _truncateThreshold = TruncateThreshold) :
        maxSize(_maxSize),
        truncateThreshold(std::max(_truncateThreshold == 0 ? _maxSize * 2 : _truncateThreshold, _maxSize))
    {
        // either specify maxSize through template arguments or the constructor and fail otherwise
        assert(_maxSize != 0);
    }
```

<sub>source: ['codex']</sub>

**Review by thepastaclaw — CHANGES_REQUESTED** [review]
## Preliminary review — Codex only

The latest delta resolves prior-1 by registering the new unit test in `test/util/data/non-backported.txt`, and it introduces no new issue. Cumulative revalidation confirms one newly identified correctness regression: explicit truncation thresholds smaller than the maximum size, which the previous implementation handled safely, can now produce an out-of-range `std::nth_element` iterator in release builds.

Validated blockers were found in the Codex precheck. Sonnet is deferred until a fresh Codex revalidation clears the blocker gate.

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — dash-core-commit-history (completed)
- Verifier: `gpt-5.6-sol` — verifier
- Sonnet: **not run (deferred by blocker gate)**

🔴 1 blocking

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/unordered_lru_cache.h`:
- [BLOCKING] src/unordered_lru_cache.h:26-33: Preserve smaller truncation thresholds
  The constructor previously accepted configurations such as `cache(10, 5)`. Although the old implementation began sorting after the sixth insertion, it did not erase anything until the size exceeded `maxSize`, so the configuration remained safe and effectively behaved as if the threshold were 10. The new assertion rejects it only in assertion-enabled builds; with `NDEBUG`, the sixth insertion reaches `std::nth_element(vec.begin(), vec.begin() + 10, vec.end(), ...)` with a six-element vector, forming an out-of-range iterator and invoking undefined behavior. Clamp the effective threshold to `maxSize` instead of relying on a debug-only assertion, and add a regression test for a threshold below the maximum.
```
</details>

File: src/unordered_lru_cache.h (hunk 1 above)
Comment (PastaPastaPasta) [reply src/unordered_lru_cache.h:34]: this is required because;
std::nth_element requires its "nth" iterator to satisfy first <= nth <= last. If truncateThreshold < maxSize, it's possible for truncate_if_needed() to be entered (cacheMap.size() > truncateThreshold) while cacheMap.size() < maxSize — meaning vec.begin() + maxSize lands past vec.end(). That's undefined behavior (out-of-bounds iterator arithmetic fed to nth_element), not just a logic bug.

Under the old std::sort-based code this misconfiguration was harmless: sorting doesn't care about maxSize at all, and the eviction loop for (i = maxSize; i < vec.size(); i++) would simply not execute if maxSize > vec.size() — a silent no-op, not memory-unsafe.


weirdly setting the truncate threshold to a max of various things makes no sense. Better to assert imo

**Comment by UdjinM6** [comment]
pls see e796c848a96688e18f0b45d1ab1cffa2ff37c78d and bccccf176a52ae4b2843d764de5dc8b535947b73

**Comment by PastaPastaPasta** [comment]
commits applied; @thepastaclaw ; please track this flake: feature_governance.py --legacy-wallet in UBSAN run

File: src/unordered_lru_cache.h [hunk 2]
```diff
@@ -29,6 +29,8 @@ class unordered_lru_cache
     {
         // either specify maxSize through template arguments or the constructor and fail otherwise
         assert(_maxSize != 0);
+        // truncate_if_needed() only runs past truncateThreshold, so this is what keeps maxSize inside the vector
+        assert(truncateThreshold >= maxSize);
```
Comment (thepastaclaw) [inline src/unordered_lru_cache.h:33]: **🟡 Suggestion: [prior-codex-1] Constructor precondition (truncateThreshold >= maxSize) enforced only via debug-only assert**

This PR's std::nth_element swap (commit 99beec4bc4e) genuinely changed the safety characteristics of misusing this constructor. Before the swap, `std::sort` plus an erase loop starting at `maxSize` meant a configuration like `cache(10, 5)` was memory-safe even though it never evicted correctly (the vector reaches size 6, the loop starting at i=10 never executes). After the swap, the same misuse reaches `std::nth_element(vec.begin(), vec.begin() + maxSize, vec.end(), ...)` with `vec.size() < maxSize`, forming an out-of-range middle iterator — undefined behavior. The follow-up commit at this head (bccccf176a5) only adds a comment explaining the precondition; it does not clamp or otherwise enforce it outside of `assert()`. Verified: no call site in `src/` currently constructs `unordered_lru_cache`/`Uint256LruHashMap` with a threshold below `maxSize` (checked all instantiations in creditpool.cpp, mnhftx.h, instantsend/db.h, llmq/*.h, saltedhasher.h). Verified: this repository's own `CMakeLists.txt` sets no explicit `CMAKE_BUILD_TYPE`/`NDEBUG` handling, so a standard `-DCMAKE_BUILD_TYPE=Release` configuration relies on CMake's built-in default flags, which do define `NDEBUG` and disable `assert()` — the risk is not purely theoretical. That said, this exactly mirrors the pre-existing, previously-unflagged `assert(nPruneAfterSize >= nMaxSize)` contract in `limitedmap.h`, and the maintainer explicitly weighed and rejected clamping in favor of a hard-fail-on-misconfiguration precondition (PR discussion, comment 3679365777). Given no live caller trips this and it matches established codebase convention, this doesn't block the PR, but it's worth clamping or adding a regression test (`cache(10, 5)` insert-past-threshold) so a future caller who mismatches the two constructor arguments fails safely in both debug and release builds rather than only in debug builds.

```suggestion
    explicit unordered_lru_cache(size_t _maxSize = MaxSize, size_t _truncateThreshold = TruncateThreshold) :
        maxSize(_maxSize),
        truncateThreshold(std::max(_truncateThreshold == 0 ? _maxSize * 2 : _truncateThreshold, _maxSize))
    {
        // either specify maxSize through template arguments or the constructor and fail otherwise
        assert(_maxSize != 0);
    }
```

<sub>source: ['claude', 'codex']</sub>

**Review by thepastaclaw — COMMENTED** [review]
## Final validation — Codex + Sonnet

This Sonnet-final pass covers the full delta cbd12002..bccccf176a5 (test coverage additions plus a doc/rename refactor of the bounded-cache prune paths) reconciled against two carried-forward prior findings. Prior-sonnet-1 (register the new unit test in non-backported.txt) is FIXED and confirmed present at the current head. Prior-codex-1 (unordered_lru_cache constructor precondition enforced only by assert, not clamped) is STILL VALID as a technical matter: independent verification confirms this PR's own nth_element commit (99beec4bc4e) changed a previously well-defined-but-degenerate edge case (truncateThreshold < maxSize) into a real out-of-bounds-iterator risk when assertions are compiled out, and this repo's CMakeLists.txt does not itself suppress CMake's default -DNDEBUG under Release builds, so the risk is real rather than purely theoretical. It is downgraded from blocking to a suggestion, however, because zero call sites in the tree currently violate the precondition, the same assert-only-precondition idiom is already an established, unflagged pattern in the adjacent limitedmap.h, and the maintainer made a considered, documented decision to assert rather than silently clamp. No new issues were introduced by the delta itself beyond this carried-forward item.

Source: Codex reviewers `gpt-5.6-sol`; Sonnet reviewers and final verifier `claude-sonnet-5`; harness `claude-code`; coordinator `openclaw-agent/cliproxy/gpt-5.6-sol` (orchestration only).

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — dash-core-commit-history (completed)
- Verifier: `claude-sonnet-5` — verifier
- Sonnet reviewers: `claude-sonnet-5` — dash-core-commit-history (completed), `claude-sonnet-5` — general (completed)

🟡 1 suggestion(s)

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `src/unordered_lru_cache.h`:
- [SUGGESTION] src/unordered_lru_cache.h:26-33: [prior-codex-1] Constructor precondition (truncateThreshold >= maxSize) enforced only via debug-only assert
  This PR's std::nth_element swap (commit 99beec4bc4e) genuinely changed the safety characteristics of misusing this constructor. Before the swap, `std::sort` plus an erase loop starting at `maxSize` meant a configuration like `cache(10, 5)` was memory-safe even though it never evicted correctly (the vector reaches size 6, the loop starting at i=10 never executes). After the swap, the same misuse reaches `std::nth_element(vec.begin(), vec.begin() + maxSize, vec.end(), ...)` with `vec.size() < maxSize`, forming an out-of-range middle iterator — undefined behavior. The follow-up commit at this head (bccccf176a5) only adds a comment explaining the precondition; it does not clamp or otherwise enforce it outside of `assert()`. Verified: no call site in `src/` currently constructs `unordered_lru_cache`/`Uint256LruHashMap` with a threshold below `maxSize` (checked all instantiations in creditpool.cpp, mnhftx.h, instantsend/db.h, llmq/*.h, saltedhasher.h). Verified: this repository's own `CMakeLists.txt` sets no explicit `CMAKE_BUILD_TYPE`/`NDEBUG` handling, so a standard `-DCMAKE_BUILD_TYPE=Release` configuration relies on CMake's built-in default flags, which do define `NDEBUG` and disable `assert()` — the risk is not purely theoretical. That said, this exactly mirrors the pre-existing, previously-unflagged `assert(nPruneAfterSize >= nMaxSize)` contract in `limitedmap.h`, and the maintainer explicitly weighed and rejected clamping in favor of a hard-fail-on-misconfiguration precondition (PR discussion, comment 3679365777). Given no live caller trips this and it matches established codebase convention, this doesn't block the PR, but it's worth clamping or adding a regression test (`cache(10, 5)` insert-past-threshold) so a future caller who mismatches the two constructor arguments fails safely in both debug and release builds rather than only in debug builds.
```
</details>

File: src/unordered_lru_cache.h (hunk 1 above)
Comment (thepastaclaw) [reply src/unordered_lru_cache.h:34]: Resolved in `bccccf1` — _Preserve smaller truncation thresholds_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

File: src/unordered_lru_cache.h (hunk 2 above)
Comment (PastaPastaPasta) [reply src/unordered_lru_cache.h:33]: asserts are not disabled in release. This assert WILL trigger in release if there is an issue. Maybe you can argue this could still open us to a DoS etc if we write buggy code for this. But the assert will trigger and we won't get to undefined behavior.

**Review by thepastaclaw — COMMENTED** [review]
## Final validation — Codex + Sonnet

Independent verification confirms the src/unordered_lru_cache.h and src/limitedmap.h nth_element pruning logic is correct, and the new/broadened unit tests in src/test/unordered_lru_cache_tests.cpp and src/test/limitedmap_tests.cpp adequately exercise truncation, recency, and threshold behavior. Carried-forward prior-codex-1 (constructor precondition truncateThreshold >= maxSize enforced only via a 'debug-only' assert) is OUTDATED: direct inspection of src/util/check.h shows `#if defined(NDEBUG) #error "Cannot compile without assertions!" #endif`, util/check.cpp is unconditionally listed in both libbitcoin_util and libbitcoinkernel in src/Makefile.am (lines 1028, 1358), and configure.ac never defines NDEBUG for the standard (non-debug) build flags -- so C assertions, including this one, are active in every supported release build of this codebase, matching the maintainer's reply (comment 3696299393) exactly. New for this delta (bccccf176a52..e843d3fad15, and confirmed within the current PR range 627f82f50a62..e843d3fad15): the final commit e843d3fad15 correctly updates src/chainlock/handler.h's sizing-rationale comment from 'sorts' to 'partitions,' but the analogous constructor-doc comment in src/limitedmap.h:34-38 (and the matching test comment in src/test/limitedmap_tests.cpp:105) still describes unordered_limitedmap::prune() as performing a full sort, even though prune() was changed in an earlier commit in this same stack to use std::nth_element (confirmed at src/limitedmap.h:109). This is a real, currently-unaddressed documentation nit kept as a nitpick finding. No backport-prerequisite claims were raised by any agent for this PR (it is an internal Dash algorithmic/perf change, not a Bitcoin Core backport), so prerequisite_adjudications is empty. Source: Sol/Codex reviewers `gpt-5.6-sol`; Sonnet reviewers and final verifier `claude-sonnet-5`; harness `claude-code`; coordinator `openclaw-agent/cliproxy/gpt-5.6-sol` (orchestration only).

### Review provenance
- Codex reviewers: `gpt-5.6-sol` — general (completed), `gpt-5.6-sol` — dash-core-commit-history (completed)
- Verifier: `claude-sonnet-5` — verifier
- Sonnet reviewers: `claude-sonnet-5` — general (failed), `claude-sonnet-5` — dash-core-commit-history (failed), `claude-sonnet-5` — general (failed), `claude-sonnet-5` — dash-core-commit-history (failed), `claude-sonnet-5` — general (completed), `claude-sonnet-5` — dash-core-commit-history (completed), `claude-sonnet-5` — backport-reviewer (completed)

💬 1 nitpick(s)

_1 additional finding(s) omitted (not in diff)._

File: src/unordered_lru_cache.h (hunk 2 above)
Comment (thepastaclaw) [reply src/unordered_lru_cache.h:33]: Resolved in this update — _[prior-codex-1] Constructor precondition (truncateThreshold >= maxSize) enforced only via debug-only assert_ no longer present.

<sub>Auto-resolved by the review system based on the latest commit diff. If you believe this was closed in error, reopen the thread.</sub>

**Comment by UdjinM6** [comment]
pls see b586d1c45e

**Comment by PastaPastaPasta** [comment]
that change should be included now; sadly the agent also did a rebase, when I wanted it to just squash down changes... range diff looks right and below is empty too. should be right.

```
git diff b586d1c45e9 b498b678c90 -- src/limitedmap.h src/
unordered_lru_cache.h src/chainlock/handler.h src/test/limite
dmap_tests.cpp src/test/unordered_lru_cache_tests.cpp test/ut
il/data/non-backported.txt
```

**Review by UdjinM6 — APPROVED** [review]
utACK b498b678c90d7f88d4bb9a6b955ce93a2652946e
